tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp

```diff
@@ -0,0 +1,164 @@
+#ifndef LINALG_TPLIMPLEMENTATIONS_INCLUDE_EXPERIMENTAL___P1673_BITS_KOKKOSKERNELS_BLAS2_MATRIX_RANK_1_UPDATE_HPP_
+#define LINALG_TPLIMPLEMENTATIONS_INCLUDE_EXPERIMENTAL___P1673_BITS_KOKKOSKERNELS_BLAS2_MATRIX_RANK_1_UPDATE_HPP_
+
+#include <complex>
+
+namespace KokkosKernelsSTD {
+
+namespace Impl {
+
+// manages parallel execution of independent action
+// called like action(i, j) for each matrix element A(i, j)
+template <typename ExecSpace, typename MatrixType>
+class ParallelMatrixVisitor {
+public:
+  KOKKOS_INLINE_FUNCTION ParallelMatrixVisitor(ExecSpace &&exec_in, MatrixType &A_in):
+    exec(exec_in), A(A_in), ext0(A.extent(0)), ext1(A.extent(1))
+  {}
+
+  template <typename ActionType>
+  void for_each_triangle_matrix_element(std::experimental::linalg::upper_triangle_t t, ActionType action) {
+    Kokkos::parallel_for(Kokkos::RangePolicy(exec, 0, ext1),
+      KOKKOS_LAMBDA(const auto j) {
+        using idx_type = std::remove_const_t<decltype(j)>;
+        for (idx_type i = 0; i <= j; ++i) {
+          action(i, j);
+        }
+      });
+    exec.fence();
+  }
+
+  template <typename ActionType>
+  void for_each_triangle_matrix_element(std::experimental::linalg::lower_triangle_t t, ActionType action) {
+    for_each_triangle_matrix_element(std::experimental::linalg::upper_triangle,
+        [action](const auto i, const auto j) {
+          action(j, i);
+      });
+  }
+
+private:
+
+  template <typename T>
+  KOKKOS_INLINE_FUNCTION
+  static bool is_int_mul_safe(T a, T b) {
+    static constexpr auto max_val = std::numeric_limits<decltype(a * b)>::max();
+    return a <= max_val / b;
+  }
+
+private:
+  const ExecSpace &exec;
```

Comment:
I've commented on other PRs on how it could be bad to store these two objects by reference, and how storing the sizes as nonconst would make the functor semiregular.

Reply:
Fixed (and discussed) on https://github.com/kokkos/stdBLAS/pull/157 (this PR is rebased on it now)